added supports for NQ processing

rongzhangibm opened the pull request:
Added:

1. scoring
primeqa/mrc/metrics/nq_f1/eval_utils.py
primeqa/mrc/metrics/nq_f1/nq_eval.py
primeqa/mrc/metrics/nq_f1/nq_f1.py

2. processors
primeqa/mrc/processors/preprocessors/natural_questions.py
primeqa/mrc/processors/postprocessors/natural_questions.py

3. unit test
tests/primeqa/mrc/metrics/nq_f1/test_nq_f1.py
(63% coverage passed.)
tests/primeqa/mrc/processors/postprocessors/test_nq_postprocessor.py
(84% coverage passed.)
tests/primeqa/mrc/processors/preprocessors/test_nq_preprocessor.py
(88% coverage passed.)

Modified:
examples/mrc/run_mrc.py
setup.py




********************
LONG ANSWER R@P TABLE:
Optimal threshold: 4.0808
 F1     /  P      /  R
 65.09% /  64.54% /  65.65%
R@P=0.5: 77.76% (actual p=50.12%, score threshold=1.979)
R@P=0.75: 41.85% (actual p=75.01%, score threshold=5.523)
R@P=0.9: 4.20% (actual p=90.32%, score threshold=8.189)
********************
SHORT ANSWER R@P TABLE:
Optimal threshold: 4.0822
 F1     /  P      /  R
 56.76% /  57.24% /  56.28%
R@P=0.5: 61.23% (actual p=50.01%, score threshold=3.235)
R@P=0.75: 29.25% (actual p=75.11%, score threshold=6.031)
R@P=0.9: 10.16% (actual p=90.00%, score threshold=7.425)
********************
METRICS IGNORING SCORES (n=7830):
              F1     /  P      /  R
Long answer   48.32% /  34.51% /  80.58%
Short answer  39.98% /  28.81% /  65.28%
***** eval metrics *****
  epoch                       =    1.0
  eval_long_answer_f1         = 0.4832
  eval_long_answer_n          =   7830
  eval_long_answer_precision  = 0.3451
  eval_long_answer_recall     = 0.8058
  eval_samples                = 289622
  eval_short_answer_f1        = 0.3998
  eval_short_answer_n         =   7830
  eval_short_answer_precision = 0.2881
  eval_short_answer_recall    = 0.6528


*******************

To run NQ:

SCRIPT=/u/zhrong/git/Public_PrimeQA/primeqa/examples/mrc/run_mrc.py
OUTPUT_DIR=/dccstor/zhrong-nmt/QA/primeqa/exp/06252022
CACHE_DIR=$OUTPUT_DIR
DATASET=/dccstor/zhrong-nmt/QA/oneqa/nq
`
python $SCRIPT  \
--model_name_or_path roberta-large \
--num_train_epochs 1 \
--output_dir $OUTPUT_DIR \
--overwrite_output_dir \
--do_train \
--do_eval \
--evaluation_strategy no \
--cache_dir $CACHE_DIR \
--per_device_train_batch_size 32 \
--per_device_eval_batch_size 64 \
--gradient_accumulation_steps 4 \
--warmup_ratio 0.1 \
--save_steps 50000 \
--learning_rate 3e-5 \
--eval_metrics NQF1 \
--postprocessor primeqa.mrc.processors.postprocessors.natural_questions.NaturalQuestionsPostProcessor \
--preprocessor primeqa.mrc.processors.preprocessors.natural_questions.NaturalQuestionsPreProcessor \
--preprocessing_num_workers 10 \
--dataset_name $DATASET  \
--dataset_config_name default \
--beam_runner DirectRunner

`

seirasto commented on examples/mrc/run_mrc.py:
```diff
@@ -197,6 +201,12 @@ class DataTrainingArguments:
+    beam_runner: str = field(
+        default=None,
+        metadata={"help": "The beam runner for loading large dataset.",
+                  "choices": 'DirectRunner',
```
Is this needed if there is only one choice right now?

rongzhangibm replied:
changed the choice of beam_runner to be a list.

seirasto commented on primeqa/mrc/processors/preprocessors/natural_questions.py:
```diff
@@ -0,0 +1,211 @@
+
+    def get_annotations(self, annotations, paragraphs):
+        """
+        Process NQ annotations into TyDi format.
```
Should this be base preprocessor format?

seirasto commented on primeqa/mrc/metrics/nq_f1/nq_eval.py:
```diff
@@ -0,0 +1,506 @@
+    return scores
+
+
+def main(predictions_path, grount_truth_file_pattern, num_workers, use_pretty_print,
```
remove main

rongzhangibm replied:
removed.

seirasto commented on primeqa/mrc/metrics/nq_f1/nq_eval.py:
```diff
@@ -0,0 +1,506 @@
+    return parser.parse_args()
+
+
+if __name__ == '__main__':
```
remove with main

rongzhangibm replied:
removed.

seirasto commented on primeqa/mrc/metrics/nq_f1/nq_f1.py:
```diff
@@ -0,0 +1,176 @@
+                    predictions={**self._common_answer_schema, **self._pred_answer_schema},
+                    references=datasets.Sequence(feature={**self._common_answer_schema, **self._ref_answer_schema})
+                )),
+            reference_urls=["https://github.com/google-research-datasets/tydiqa/blob/master/tydi_eval.py"],
```
wrong reference?

rongzhangibm replied:
fixed.

seirasto commented on primeqa/mrc/metrics/nq_f1/nq_f1.py:
```diff
@@ -0,0 +1,176 @@
+            short_non_null_threshold=short_non_null_threshold)
+
+        metrics = pretty_print(long_answer_stats=long_answer_stats, short_answer_stats=short_answer_stats)
+        # metrics = get_metrics_with_answer_stats(long_answer_stats, short_answer_stats)
```
remove commented line

rongzhangibm replied:
removed.

seirasto reviewed (commented):
Please see my comments. Thanks!

rongzhangibm commented on primeqa/mrc/metrics/nq_f1/nq_f1.py:
```diff
@@ -0,0 +1,176 @@
+from primeqa.mrc.metrics.nq_f1.nq_eval import pretty_print, get_metrics_with_answer_stats, score_answers
+from primeqa.mrc.data_models.target_type import TargetType
+
+# TODO: NQ F1 doc
```
removed

rongzhangibm commented on tests/primeqa/mrc/metrics/nq_f1/test_nq_f1.py:
```diff
@@ -0,0 +1,70 @@
+            [dict(start_position=-1, end_position=-1, passage_index=-1,
+                  yes_no_answer=TargetType.NO_ANSWER, example_id='9484', language='english')] +
+            [dict(start_position=37907, end_position=37946, passage_index=1,
+                  yes_no_answer=TargetType.NO, example_id='9484', language='english')] * 2 +
```
fixed.

rongzhangibm commented on examples/mrc/run_mrc.py:
```diff
@@ -352,11 +362,20 @@ def main():
+                data_args.dataset_name,
+                data_args.dataset_config_name,
+                cache_dir=model_args.cache_dir,
+                beam_runner=data_args.beam_runner
```
If beam_runner is NONE, the following code (from line 373) will be performed which doesn't use beam_runner to load dataset.
`            raw_datasets = datasets.load_dataset(
                data_args.dataset_name,
                data_args.dataset_config_name,
                cache_dir=model_args.cache_dir
            )
`

seirasto reviewed (commented):
LGTM. I think just the run_mrc README needs to be updated to include that NQ is available. Also, do we need a comment for how to run now with `load_dataset("natural_questions", revision="main")`?

bhavani105 commented on primeqa/mrc/processors/preprocessors/natural_questions.py:
```diff
@@ -0,0 +1,233 @@
+             Annotations in preprocessor format.
+        """
+
+        tydi_annotations = {}
```
Would be better to call this nq_annotations

rongzhangibm replied:
done

bhavani105 reviewed (commented):
LGTM
Minor changes
mrc/README.md has to be updated with how to run NQ, expected results, resources required and any hints on how to run with a large dataset.

rongzhangibm commented:
> 

Added nq steps to mrc/README.md.
Added revision="main" to nq dataset loading in run_mrc.py. Once HF release the new natural_questions dataset, we can remove it with a new PR.

bhavani105 commented on primeqa/mrc/metrics/nq_f1/nq_f1.py:
```diff
@@ -0,0 +1,174 @@
+F1 = 2 * (precision * recall) / (precision + recall).  This implementation of F1 is based
+on the TyDi QA leaderboard.
+
+Adapted from https://github.com/google-research-datasets/tydiqa/blob/master/tydi_eval.py.
```
TyDIQA references - pls update

bhavani105 commented on primeqa/mrc/metrics/nq_f1/nq_f1.py:
```diff
@@ -0,0 +1,174 @@
+year    = {2020},
+journal = {Transactions of the Association for Computational Linguistics}
+}
+"""
```
TyDIQA references - pls update

rongzhangibm replied:
done

bhavani105 reviewed (approved):
LTGM

Please update the citations

seirasto reviewed (approved):
LGTM!

jsmcibm reviewed (approved):
The inclusion of beam runner forces to downgrade pyarrow to version 7.0.0.   This overrides the pyarrow downgrade proposed in [298](https://github.com/primeqa/primeqa/pull/298), which should probably be closed.
I have verified that a fresh condo environment the boolean extensions are still working.

I worry that there is a lot of duplicate code between tydi metrics and nq metrics - if this is code that comes from the tydi and nq distributions, we should probably maintain it as separate, but with known issues like the hard-coded languages in the tydi metric, a merged metric may become necessary in the future.

Otherwise LGTM
